### ISSUE TITLE
Emojis in grammar cause assertion failure

### ISSUE THREAD

**Issue by ai-and-i** [ggerganov/llama.cpp#2501]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [x] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [x] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Expected Behavior

I'm trying to make a model respond only using emojis. I'm using the following grammar-file for it:

```
root ::= emoji+
emoji ::= [😀-🙏]
```

EDIT: I also tried unicode escapes, and a simpler grammar (see bellow), but the result is the same:

```
root ::= "\U0001F600"
```

# Current Behavior

llama.cpp crashes with assertion failure:

```
### Assistant:LLAMA_ASSERT: /Users/test/llama.cpp/llama.cpp:2873: !grammar->stacks.empty()
[1]    56512 abort      ./bin/main -m ../models/stablebeluga-7b.ggmlv3.q4_K_M.bin -f ./prompt.txt -c 102
```

# Environment and Context

This is M1 macbook pro with macos 13.4.1.

# Steps to Reproduce

1. create grammar.gbnf with the content below
2. launch llama.cpp with the above grammar file

# Environment info:
```
llama.cpp$ git log | head -1
commit 8183159cf3def112f6d1fe94815fce70e1bffa12

llama.cpp$ make --version | head -1
GNU Make 3.81

llama.cpp$ g++ --version
Apple clang version 14.0.3 (clang-1403.0.22.14.1)
Target: arm64-apple-darwin22.5.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Comment by ai-and-i** [comment]
Looking deeper into this, the issue is quite fundamental: the grammar system works in unicode and assumes that each token produces one or more full unicode characters. However, that assumption is wrong — tokenizers work on bytes rather than characters, and many unicode characters are encoded into multiple tokens. E.g., `😀` is encoded as 4 tokens: `243, 162, 155, 131`. Decoding each of these tokens alone produces invalid utf-8 sequences.

I think the best solution is to switch the grammar system to work on byte level. It would mean that ranges like `[😀-🙏]` or even `[α-δ]` would need to be transformed into something like `[\xce] [\xb1-\xb4]` sequences internally.

**Comment by ejones** [comment]
Ugh, yeah you're right. I should've realized that. I was hoping to avoid rewriting ranges but I guess it's necessar. Wondering if I can maybe get away with just tracking a partial UTF-8 sequence alongside the grammar stack.

**Comment by ai-and-i** [comment]
Another approach I thought of would be to convert incomplete characters to ranges of unicode codepoints. E.g., `\xce` can be completed to anything from b`\xce\x80` to b`\xce\xbf`, which is [\U00000380-\U000003bf]. When evaluating, you can then check for overlap of this range and the range allowed by the grammar. However that would require keeping the current index inside the incomplete sequence being generated — so in this example, when you get the next byte `\xb1`, you'd need to have the previous incomplete bytes around to decode it. To me, it seems more complicated than just rewriting the ranges.

**Comment by ejones** [comment]
I like the partial decoding approach actually. Hoping to find time to work on it tonight.

**Comment by ejones** [comment]
Put up a draft fix in #2553. Plan to come back and flesh out/clean it up in a bit.
